<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
!pip install psycopg2

In [3]:
connection = psycopg2.connect(
    dbname = 'project_sql',
    user = 'skillfactory',
    password = 'cCkxxLVrDE8EbvjueeMedPKt',
    host = '84.201.134.129',
    port = 5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [4]:
# текст запроса
query_3_1 = f'''
    SELECT COUNT(id) AS vacancies_count
    FROM public.vacancies
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_18213/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


,vacancies_count
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [6]:
# текст запроса
query_3_2 = f'''
    SELECT COUNT(id) AS employers_count
    FROM public.employers
'''

In [7]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

/tmp/ipykernel_18213/533540996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


,employers_count
0,23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
    SELECT COUNT(id) AS areas_count
    FROM public.areas
'''

In [9]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

/tmp/ipykernel_18213/1395976625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


,areas_count
0,1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
    SELECT COUNT(id) AS industries_count
    FROM public.industries
'''

In [11]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

/tmp/ipykernel_18213/854197215.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


,industries_count
0,294


***

Масштаб датасета:

В нашем распоряжении находится достаточно крупная база данных, включающая 49 197 вакансий.

Разнообразие рынка:

 Вакансии представлены 23 501 уникальным работодателем. В среднем на одну компанию приходится около двух вакансий. Это отличный показатель того, что выборка разнообразна и объективно отражает текущую ситуацию на рынке труда.

Широкая география:

 Данные охватывают 1 362 различных региона. Такая географическая распределенность позволит нашей будущей модели учитывать региональную специфику.

Многообразие сфер деятельности:

 Присутствие в базе 294 различных сфер деятельности показывает, что специалисты требуются в совершенно разных отраслях бизнеса.

Общий итог:

 База данных является репрезентативной, масштабной и детализированной. Структура таблиц и полнота справочников позволяют переходить к следующему этапу — более глубокому и детальному анализу самих вакансий.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
# текст запроса
query_4_1 = f'''
    SELECT
        a.name AS area,
        COUNT(v.id) AS cnt
    FROM public.vacancies v
    JOIN public.areas a ON v.area_id = a.id
    GROUP BY a.name
    ORDER BY cnt DESC
'''

In [13]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

/tmp/ipykernel_18213/620650199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


,area,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
...,...,...
764,Тарко-Сале,1
765,Новоаннинский,1
766,Бирск,1
767,Сасово,1


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
    SELECT COUNT(id) AS vacancies_count
    FROM public.vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [15]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

/tmp/ipykernel_18213/3746829199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


,vacancies_count
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
    SELECT
        ROUND(AVG(salary_from)) AS avg_salary_from,
        ROUND(AVG(salary_to)) AS avg_salary_to
    FROM public.vacancies
'''

In [17]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

/tmp/ipykernel_18213/3595259731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
    SELECT
        schedule,
        employment,
        COUNT(id) AS cnt
    FROM public.vacancies
    GROUP BY schedule, employment
    ORDER BY cnt DESC
'''

In [19]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

/tmp/ipykernel_18213/2028972612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


,schedule,employment,cnt
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [20]:
# текст запроса
query_4_5 = f'''
    SELECT
        experience,
        COUNT(id) AS cnt
    FROM public.vacancies
    GROUP BY experience
    ORDER BY cnt ASC
'''

In [21]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

/tmp/ipykernel_18213/210463167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


,experience,cnt
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

География вакансий:

Рынок труда имеет ярко выраженную централизацию. Безоговорочным лидером по числу вакансий является Москва (5 333), за ней с заметным отставанием идут: Санкт-Петербург (2 851), Минск (2 112) и Новосибирск (2 006). Нашей рекомендательной системе определенно придется учитывать локацию пользователя как один из важнейших факторов.

Проблема пропусков в данных о зарплате:

 Из 49 197 вакансий хотя бы одна из границ зарплатной вилки указана лишь в 24 073 случаях. Это означает, что около половины (51%) работодателей скрывают уровень дохода. С этим придется что-то делать.

Средний уровень дохода:

В тех вакансиях, где доход указан, средняя нижняя граница составляет около 71 065 рублей, а верхняя — 110 537 рублей. Это дает нам базовый ориентир рынка

Популярные форматы работы:

 Абсолютный стандарт рынка — это «Полный день / Полная занятость» (35 367 вакансий). На второе место попадает «Удаленная работа / Полная занятость» (7 802 вакансии).

Требования к опыту:

Наиболее востребованы на рынке специалисты уровня Middle (от 1 года до 3 лет — 26 152 вакансии) и Middle/Senior (от 3 до 6 лет — 14 511 вакансий). Вакансий для начинающих без опыта (Junior) заметно меньше — 7 197, но рынок для них все равно существует. Вакансий для специалистов с опытом более 6 лет (Lead/Principal) меньше всего (1 337).

Общий итог этапа:

Мы увидели, что данные неоднородны. Есть выраженные лидеры по городам и форматам работы, а также проблема с пропусками в зарплатах.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
    SELECT
        e.name AS employer_name,
        COUNT(v.id) AS vacancies_count
    FROM public.employers e
    JOIN public.vacancies v ON e.id = v.employer_id
    GROUP BY e.id, e.name
    ORDER BY vacancies_count DESC
    LIMIT 5
'''

In [23]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

/tmp/ipykernel_18213/1147638897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


,employer_name,vacancies_count
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''
    WITH vac_counts AS (
        SELECT area_id, COUNT(id) AS vacancies_count
        FROM public.vacancies
        GROUP BY area_id
    ),
    emp_counts AS (
        SELECT area, COUNT(id) AS employers_count
        FROM public.employers
        GROUP BY area
    )

    SELECT
        a.name AS area_name,
        COALESCE(v.vacancies_count, 0) AS vacancies_count,
        COALESCE(e.employers_count, 0) AS employers_count
    FROM public.areas a
    LEFT JOIN vac_counts v ON a.id = v.area_id
    LEFT JOIN emp_counts e ON a.id = e.area
    ORDER BY vacancies_count ASC, employers_count DESC
'''

In [25]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2.head(10) # Выведем топ-10, чтобы сразу увидеть нужный регион с нулями вакансий

/tmp/ipykernel_18213/58021675.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


,area_name,vacancies_count,employers_count
0,Россия,0,410
1,Казахстан,0,207
2,Московская область,0,75
3,Краснодарский край,0,19
4,Ростовская область,0,18
5,Беларусь,0,18
6,Азербайджан,0,17
7,Нижегородская область,0,16
8,Республика Татарстан,0,16
9,Узбекистан,0,15


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''
    SELECT
        e.name AS employer_name,
        COUNT(DISTINCT v.area_id) AS regions_count
    FROM public.employers e
    JOIN public.vacancies v ON e.id = v.employer_id
    GROUP BY e.id, e.name
    ORDER BY regions_count DESC
'''

In [27]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

/tmp/ipykernel_18213/133400711.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


,employer_name,regions_count
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
14901,НПП Авиатрон,1
14902,Центр дистанционных торгов,1
14903,Городские Телекоммуникационные Системы,1
14904,"Введенский, Отель",1


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [28]:
# текст запроса
query_5_4 = f'''
    SELECT COUNT(e.id) AS employers_without_industry
    FROM public.employers e
    LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
    WHERE ei.industry_id IS NULL
'''

In [29]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

/tmp/ipykernel_18213/1045119256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


,employers_without_industry
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [30]:
# текст запроса
query_5_5 = f'''
    SELECT e.name AS employer_name
    FROM public.employers e
    JOIN public.employers_industries ei ON e.id = ei.employer_id
    GROUP BY e.id, e.name
    HAVING COUNT(ei.industry_id) = 4
    ORDER BY e.name ASC
    OFFSET 2 LIMIT 1
'''

In [31]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

/tmp/ipykernel_18213/4025176131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


,employer_name
0,2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
    SELECT COUNT(ei.employer_id) AS employers_count
    FROM public.employers_industries ei
    JOIN public.industries i ON ei.industry_id = i.id
    WHERE i.name = 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

/tmp/ipykernel_18213/2713946032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


,employers_count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [34]:
# код для получения списка городов-милионников
million_cities = (
    'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань',
    'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа',
    'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград'
)


In [35]:
# текст запроса
query_5_7 = f'''
    WITH yandex_vacs AS (
        SELECT
            a.name AS region,
            COUNT(v.id) AS cnt
        FROM public.vacancies v
        JOIN public.employers e ON v.employer_id = e.id
        JOIN public.areas a ON v.area_id = a.id
        WHERE e.name = 'Яндекс' AND a.name IN {million_cities}
        GROUP BY a.name
    )
    SELECT region, cnt FROM yandex_vacs
    UNION ALL
    SELECT 'Total', SUM(cnt) FROM yandex_vacs
    ORDER BY cnt ASC
'''

In [36]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

/tmp/ipykernel_18213/2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


,region,cnt
0,Омск,21.0
1,Челябинск,22.0
2,Красноярск,23.0
3,Волгоград,24.0
4,Пермь,25.0
5,Казань,25.0
6,Ростов-на-Дону,25.0
7,Самара,26.0
8,Уфа,26.0
9,Краснодар,30.0


***

Лидеры рынка:

 Абсолютным монополистом на рынке найма является «Яндекс» с 1 933 вакансиями. За ним следуют Ростелеком, Тинькофф, СБЕР. Это говорит о том, что основной спрос на IT-кадры формируется крупным бизнесом.

Широта охвата:

 Тот же «Яндекс» лидирует и по географическому охвату, публикуя вакансии в 181 регионе. «Ростелеком» охватывает 152 региона.

Специфика заполнения географии:

 наблюдается интересная особенность архитектуры данных: 410 работодателей зарегистрированы в регионе «Россия», но на этот макрорегион не заведено ни одной вакансии. Это связано с тем, что работодатель может указать в профиле страну, но сами вакансии всегда привязываются к конкретным городам.

Пропуски в данных:

Огромное число работодателей — 8 419 не указали сферу своей деятельности. Для нашей ML-модели это означает наличие большого кластера «неизвестных» индустрий, с которым придется работать

Доля чистого IT:

 В сфере "Разработка программного обеспечения" работает 3 553 компании. Это солидное ядро профильных IT-компаний, однако мы видим, что Data-специалисты требуются не только в IT, но и в тысячах других бизнесов.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [52]:
# текст запроса
query_6_1 = f'''
    SELECT COUNT(id) AS data_vacancies_count
    FROM public.vacancies
    WHERE name ILIKE '%data%' OR name ILIKE '%данн%'
'''

In [53]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

/tmp/ipykernel_18213/1281472032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


,data_vacancies_count
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [54]:
# текст запроса
query_6_2 = f'''
    SELECT COUNT(id) AS junior_ds_count
    FROM public.vacancies
    WHERE (
        name ILIKE '%data scientist%' OR
        name ILIKE '%data science%' OR
        name ILIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
        name ILIKE '%machine learning%' OR
        name ILIKE '%машинн%обучен%'
    )
    AND (
        name ILIKE '%junior%' OR
        experience = 'Нет опыта' OR
        employment = 'Стажировка'
    )
'''

In [55]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

/tmp/ipykernel_18213/118162426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


,junior_ds_count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [56]:
# текст запроса
query_6_3 = f'''
    SELECT COUNT(id) AS sql_postgres_ds_count
    FROM public.vacancies
    WHERE (
        name ILIKE '%data scientist%' OR
        name ILIKE '%data science%' OR
        name ILIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
        name ILIKE '%machine learning%' OR
        name ILIKE '%машинн%обучен%'
    )
    AND (key_skills ILIKE '%sql%' OR key_skills ILIKE '%postgres%')
'''

In [57]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

/tmp/ipykernel_18213/3310962512.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


,sql_postgres_ds_count
0,201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [58]:
# текст запроса
query_6_4 = f'''
    SELECT COUNT(id) AS python_ds_count
    FROM public.vacancies
    WHERE (
        name ILIKE '%data scientist%' OR
        name ILIKE '%data science%' OR
        name ILIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
        name ILIKE '%machine learning%' OR
        name ILIKE '%машинн%обучен%'
    )
    AND key_skills ILIKE '%python%'
'''

In [59]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

/tmp/ipykernel_18213/614304167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


,python_ds_count
0,351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [61]:
# текст запроса
query_6_5 = f'''
    SELECT ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) AS avg_skills_count
    FROM public.vacancies
    WHERE (
        name ILIKE '%data scientist%' OR
        name ILIKE '%data science%' OR
        name ILIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
        name ILIKE '%machine learning%' OR
        name ILIKE '%машинн%обучен%'
    )
    AND key_skills IS NOT NULL
'''

In [62]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

/tmp/ipykernel_18213/4089566367.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


,avg_skills_count
0,6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [63]:
# текст запроса
query_6_6 = f'''
    SELECT
        experience,
        ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to))) AS avg_salary
    FROM public.vacancies
    WHERE (
        name ILIKE '%data scientist%' OR
        name ILIKE '%data science%' OR
        name ILIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
        name ILIKE '%machine learning%' OR
        name ILIKE '%машинн%обучен%'
    )
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
    GROUP BY experience
'''

In [64]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

/tmp/ipykernel_18213/2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


,experience,avg_salary
0,Нет опыта,74643.0
1,От 1 года до 3 лет,139675.0
2,От 3 до 6 лет,243115.0


***

Выводы по предметному анализу:

Объем рынка данных:

 Из почти 50 тысяч вакансий в базе только 1 771 имеет прямое отношение к данным. Это подтверждает, что работа с данными — это узкоспециализированная ниша на общем фоне рынка труда.

Высокий порог входа:

 Для начинающих специалистов (Junior) открыта всего 51 вакансия (это менее 3% от всех вакансий, связанных с данными). Рынок Data Science остро нуждается в специалистах уровня Middle и Senior, но неохотно берет новичков. Джунам предстоит жесткая конкуренция :((

Ключевые навыки:

 Python ожидаемо является абсолютным маст-хэвом — он указан в 351 вакансии, в то время как SQL/Postgres требуется в 201 случае. Среднее количество требуемых навыков (6.41) говорит о том, что от дата-сайентиста ждут владения целым стеком технологий, а не только одним языком программирования.

Финансовая мотивация и рост:

 Зарплатная вилка демонстрирует впечатляющий рост в зависимости от опыта. На старте («Нет опыта») специалисты могут рассчитывать в среднем на 74 643 руб. Однако уже через 1–3 года средняя зарплата взлетает почти в два раза (до 139 675 руб.), а специалисты с опытом от 3 до 6 лет получают в среднем 243 115 руб. Это компенсирует высокий порог входа в профессию.

# Общий вывод по проекту

подведем итог исследования, обобщите выводы

здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

В ходе работы над проектом была проанализированна база данных HeadHunter. Рынок IT-вакансий сильно централизован вокруг крупнейших городов (Москва, Санкт-Петербург) и крупных корпораций (Яндекс, Ростелеком, СБЕР). При этом активно развивается формат удаленной работы, что стирает географические границы для соискателей.

Что касается непосредственно Data Science — это высокооплачиваемая, требовательная к навыкам (в среднем 6-7 ключевых компетенций, база — Python и SQL) и сложная для старта ниша. Основной спрос и максимальный скачок в зарплатах начинаются с отметки от 1 года опыта.

In [66]:
# Дополнительное исследование: Кластеризация DS-вакансий
query_ = f'''
    WITH ds_vacancies AS (
        SELECT id, name
        FROM public.vacancies
        WHERE (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
    )
    SELECT
        CASE
            WHEN name ILIKE '%computer vision%' OR name LIKE '%CV%' OR name ILIKE '%компьютерн%зрен%' THEN 'Computer Vision (CV)'
            WHEN name ILIKE '%nlp%' OR name ILIKE '%natural language%' OR name ILIKE '%текст%' THEN 'Natural Language Processing (NLP)'
            WHEN name ILIKE '%analyst%' OR name ILIKE '%аналитик%' THEN 'Data Analyst'
            WHEN name ILIKE '%engineer%' OR name ILIKE '%инженер%' OR name ILIKE '%разработчик%' THEN 'ML / Data Engineer'
            ELSE 'General Data Scientist / Researcher'
        END AS specialization,
        COUNT(id) AS count
    FROM ds_vacancies
    GROUP BY specialization
    ORDER BY count DESC
'''

# результат запроса
df_ = pd.read_sql_query(query_, connection)
df_

/tmp/ipykernel_18213/2430242521.py:30: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_ = pd.read_sql_query(query_, connection)


,specialization,count
0,General Data Scientist / Researcher,280
1,ML / Data Engineer,120
2,Data Analyst,48
3,Computer Vision (CV),19
4,Natural Language Processing (NLP),13


### Дополнительное исследование: Анализ узких специализаций

Чтобы сделать нашу рекомендательную систему более точной, было проведено дополнительное исследование: кластеризация вакансий в сфере Data Science по узким специализациям с помощью SQL-оператора CASE WHEN.

Понятие "Data Scientist" на рынке труда сильно размыто. Значительная часть вакансий фактически ищет ML/Data инженеров или Аналитиков данных, хотя они и попадают под широкие фильтры DS.

Мы можем выделить четкие ниши: специалисты по компьютерному зрению и обработке естественного языка - NLP.

Ценность для проекта:

 Наша ML-модель не должна рекомендовать все DS-вакансии подряд. Если соискатель специализируется на компьютерном зрении, ему нет смысла предлагать классическую аналитику, и наоборот. Внедрение признака «Специализация»  на основе анализа текста названий вакансий значительно повысит качество  рекомендаций для клиентов агентства